### ISSUE TITLE
Ability to download zip, tar etc?

### ISSUE THREAD

**Issue by shirshak55** [joseluisq/static-web-server#67]
As I store directories it would be nice if we can download the whole folder using zip or tar. I think it's a standard functionality for a static web server?

**Comment by joseluisq** [comment]
Why not but one question, can you elaborate a bit more about how this feature will behave?

Or similar to https://www.nginx.com/resources/wiki/modules/zip/ maybe?
I think we want something more "easy" than that one or?

**Comment by shirshak55** [comment]
@joseluisq thanks for the concern

I think we compress the directory (risk for using too much disk if not cleaned up) using any zip or tar crate and serve it. And I have seen many apps which directly compress in memory (this has a risk of oom).

You can check h5ai code here.

TAR: https://github.com/lrsjng/h5ai/blob/aa94de4945e69ff742525cd4d648c91e5c9d8cd2/src/_h5ai/private/php/ext/class-archive.php#L94

We can use any techniques afaik

**Comment by joseluisq** [comment]
Ok, another thing. If I understood correctly. Whether you would be able to enable the feature then all directories will be able (forced) to download only? How should this behave along with the directory listing feature? Ideas?

**Comment by Fukitsu** [comment]
I think he probably means something like https://github.com/svenstaro/miniserve option to download a folder or the whole thing as tar or zip

**Comment by joseluisq** [comment]
@Fukitsu Thanks for clarifying. Then it could be an interesting feature.

**Comment by palant** [comment]
This sounds like something that would be tied to the `directory-listing` feature. At the very least, the download links should be exposed in the directory listings.

Just a note: there is absolutely no problem generating a TAR archive on the fly. ZIP format on the other hand isn’t great for streaming, and most implementations (including the one used by miniserv) would require caching the entire ZIP file in memory before it can be sent. It looks like [archflow](https://github.com/plrigaux/archflow) is the only usable implementation right now that allows streaming ZIP files directly. Disclaimer says: “The library is in construction and not stable. More tests need to be done.”

**Comment by ekangmonyet** [comment]
Hi, I have a working streaming prototype for this and is interested in polishing the changes and ironing out some design choices and implementation details. I will create a PR soon, after some basic tidying.

EDIT: Listing some details here that need discussion/inputs from yall, currently:

- add a "virtual file" for directory, e.g. if directory is `/test-dir/`, the archive can be downloaded from `/test-dir/.tar.gz`
- when listing directory, the "virtual file" is labeled as `(download as .tar.gz)`
- when handling `/test-dir/.tar.gz`, an async task is spawned, and use `async-tar` and existing `GzipEncoder` to write to an mpsc channel `Body::channel()`
- use content disposition header to hint browser

I suppose the `.tar.gz` suffix and virtual file might need to be configurable/handled using other method, and the mpsc capacity should also be configurable to limit memory usage? Should other compression methods be supported as well? (by using `async-compression` it should be trivial to also support XZ and bzip2, which probably make up the three most common compressed tarballs...?)

**Comment by joseluisq** [comment]
> Hi, I have a working streaming prototype for this and is interested in polishing the changes and ironing out some design choices and implementation details. I will create a PR soon, after some basic tidying.
> 
> EDIT: Listing some details here that need discussion/inputs from yall, currently:
> 
> * add a "virtual file" for directory, e.g. if directory is `/test-dir/`, the archive can be downloaded from `/test-dir/.tar.gz`
> * when listing directory, the "virtual file" is labeled as `(download as .tar.gz)`
> * when handling `/test-dir/.tar.gz`, an async task is spawned, and use `async-tar` and existing `GzipEncoder` to write to an mpsc channel `Body::channel()`
> * use content disposition header to hint browser

@ekangmonyet It sounds fine to me, feel free to open a draft PR.

> I suppose the `.tar.gz` suffix and virtual file might need to be configurable/handled using other method, and the mpsc capacity should also be configurable to limit memory usage? Should other compression methods be supported as well? (by using `async-compression` it should be trivial to also support XZ and bzip2, which probably make up the three most common compressed tarballs...?)

Certainly! I believe what we want is to see how the feature might work at first. If you ask me, I would go with `.tar.gz`-only for now. Then we could refine/enhance the details later.

**Comment by joseluisq** [comment]
Closing as it's implemented by #544 (tar.gz).

**Comment by yonas** [comment]
@joseluisq Thanks!

Can we add support for zstd and zip?

**Comment by joseluisq** [comment]
> @joseluisq Thanks!
> 
> Can we add support for zstd and zip?

For now, we are going with tar.gz but I believe we can support those two in the future.

cc' @ekangmonyet if has thoughts in that matter.

**Comment by ekangmonyet** [comment]
> Can we add support for zstd and zip?

zstd seems to be trivial to implement, since it can be implemented using `async_compression` too, which is what we are already using right now.  I am not familiar with that algo but from what I gathered (in a short timeframe tbh), if you (or someone else) have requirement for this specific format instead of the conventional gunzip, will they also have specific requirements on how it is configured (e.g. https://docs.rs/async-compression/latest/async_compression/zstd/struct.CParameter.html)?

Regarding zip: as mentioned above, [archflow](https://github.com/plrigaux/archflow) is still a work in progress. I understand that zip format will be very useful as it is supported natively by Windows, but that might also be the exact reason why we should be more conservative when pulling that library in, the impact radius might be too risky?

To summarize, for zip to come, it's either:
 
- it is acceptable to experiment with **archflow**, or
- **archflow** is out of alpha (or whatever state we deemed too unstable), or
- it is acceptable to use/explore a non-streaming solution, or
- a stable streaming zip library is available

**Comment by joseluisq** [comment]
@ekangmonyet thanks for the insights.

About zip, since archflow is still not stable I do not believe we can use it yet as an stable feature in SWS yet (maybe as experimental?). However, As far as I can see, [Windows 11 already has support for reading several archive formats like `tar.gz` or ` tar.zst` natively since 2023](https://blogs.windows.com/windows-insider/2023/08/18/announcing-windows-11-insider-preview-build-22621-2199-and-22631-2199-beta-channel/). So we could wait for zip a bit.

About zstd, I agree, it can be implemented via the `async_compression` crate (already part of SWS dependency tree). But I guess we have to use `tar.zst` like similar to `tar.gz`.
So, if you ask me, I would like to see zstd be supported in SWS in the near future than zip.

**Comment by ekangmonyet** [comment]
Ah, didn't know Windows supported them, that's good to know.

I can help explore supporting zstd, but note that it might not be that timely. 

Do we want to open another issue specifically for zstd for tracking purpose?

**Comment by joseluisq** [comment]
> I can help explore supporting zstd, but note that it might not be that timely.

@ekangmonyet Sure, take your time.
 
> Do we want to open another issue specifically for zstd for tracking purpose?

Yes, it should be better.
